### PR TITLE
NT-0013 - Added User Model + register route + (NT-0002 task)

### PR DESCRIPTION
**Pull request by kamilmoskal**
I did this register user route similar like in my other app which i am creating. What do you think about it?
Password is encrypted (bcryptjs). After success registration in response user getting jwt token which is valid 1hour.
Each request body is validating (express-validator) and if is not valid in response we get specific error msg.

### CODE REVIEW

File: controllers/users/usersValidator.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+const { check } = require("express-validator/check");
+
+exports.validate = method => {
+  switch (method) {
+    case "registerUser": {
+      return [
```
Comment (mras2303) [inline controllers/users/usersValidator.js]: To avoid typo I suggest move ```registerUser``` to constant

File: controllers/users/users.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      .json({ errors: errors.array({ onlyFirstError: true }) });
+  }
+
+  const { firstName, lastName, email, password } = req.body;
+
+  User.findOne({ email: email.toLowerCase() }).then(user => {
+    if (user) {
+      return res
+        .status(400)
+        .json({ errors: [{ msg: "Email already exists" }] });
```
Comment (mras2303) [inline controllers/users/users.js]: To avoid future messages move it to another file with all messages

File: controllers/users/users.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    } else {
+      const newUser = new User({
```
Comment (damian0o) [inline controllers/users/users.js]: Why do we need to create another instance of user object ? Can't we pass just req.body ?

File: controllers/users/users.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+              jwt.sign(
+                payload,
+                config.get("jwtSecret"),
+                {
+                  expiresIn: 3600
+                },
+                (err, token) => {
+                  if (err) throw err;
+                  res.json({ msg: "User registered", token });
```
Comment (damian0o) [inline controllers/users/users.js]: Why do we want to send some jwt token as response after registration ?

File: controllers/users/users.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        .status(400)
+        .json({ errors: [{ msg: "Email already exists" }] });
+    } else {
+      const newUser = new User({
+        firstName,
+        lastName,
+        email,
+        password
+      });
+      bcrypt.genSalt(10, (err, salt) => {
```
[the remark this answers is not in the document]
Comment (kamilmoskal) [reply controllers/users/users.js]: In login route we can use simply one line to check if password is same as in db: bcrypt.compare(req.body.password, user.password).then(isMatch =>...
idk exactly how it works too but it hashing and ensures some security in our db

File: controllers/users/users.js (hunk 4 above)
Comment (kamilmoskal) [reply controllers/users/users.js]: to log in the user right away after registration success, instead of register and then login

File: controllers/users/users.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (damian0o) [reply controllers/users/users.js]: If you are sure that this one will work i am ok with your answer.

File: controllers/users/users.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+  if (!errors.isEmpty()) {
+    return res
+      .status(400)
+      .json({ errors: errors.array({ onlyFirstError: true }) });
+  }
+
+  const { firstName, lastName, email, password } = req.body;
+
+  User.findOne({ email: email.toLowerCase() }).then(user => {
```
Comment (mras2303) [inline controllers/users/users.js]: I'm afraid of finding users by toLowerCase email.
What if user create account with uppercase letters? There was no replacing to lowercase new email in your method.
'EXAMPLE@gmail.com' (DB) != 'example@gmail.com' (FINDING USER)

So theoriticaly user can create many of account with different letters lower or upper

Correct me if I'm wrong

Comment (kamilmoskal) [reply controllers/users/users.js]: You're right.

File: controllers/users/users.js (hunk 3 above)
Comment (kamilmoskal) [reply controllers/users/users.js]: I'm not sure if i am uderstanding what do you asking. We need(?) to pass data through new User model to save user in users collection,

File: controllers/users/users.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+              jwt.sign(
+                payload,
+                config.get("jwtSecret"),
+                {
+                  expiresIn: 3600
+                },
+                (err, token) => {
+                  if (err) throw err;
+                  res.json({ msg: "User registered", token });
+                }
```
[the remark this answers is not in the document]
Comment (mras2303) [reply controllers/users/users.js]: UR say nothing to me. Please rephrase it to USER_REGISTERED_MSG or something like that

File: controllers/users/users.js [hunk 8]
```diff
@@ -17,14 +18,12 @@ exports.registerUser = (req, res) => {
 
   User.findOne({ email: email.toLowerCase() }).then(user => {
     if (user) {
-      return res
-        .status(400)
-        .json({ errors: [{ msg: "Email already exists" }] });
+      return res.status(400).json({ errors: [{ msg: EAE }] });
```
Comment (mras2303) [inline controllers/users/users.js]: EAE say nothing to me. Please rephrase it to EMAIL_ALREADY_EXISTS or something like that

File: controllers/users/users.js [hunk 9]
```diff
@@ -3,6 +3,7 @@ const bcrypt = require("bcryptjs");
 const jwt = require("jsonwebtoken");
 const config = require("config");
 const User = require("../../models/User");
+const { SE, EAE, UR } = require("../messages");
```
Comment (mras2303) [inline controllers/users/users.js]:
```
const messages = require('../messages')
```

It won't be better?
And in every messages changing to -> message.USER_DOESNT_EXIST
